### DISCUSSION TITLE
Is file cancellation supported?

### DISCUSSION THREAD

**Discussion by braineo** [emacs-lsp/lsp-mode#3127] [Q&A] [answered]
Hi!

For pyright, it seems the client in vscode supports file based cancellation for fast request cancelling. I'm checking lsp-mode but I did not manage to find anything related. It seems only json-rpc is supported?

for file based cancellation, the pyright process needs another argument as discussed. and vscode extension code is [here](https://github.com/microsoft/pyright/tree/master/packages/vscode-pyright). Maybe need to mimic the behavior as

1. when vsc extension is activated, `cancellationStrategy` is instantiated

https://github.com/microsoft/pyright/blob/7bc2bd8ab05fc4967120d1b6b2a0a1834da2b783/packages/vscode-pyright/src/extension.ts#L64
```ts
cancellationStrategy = new FileBasedCancellationStrategy();
```

2. it pick up a random name, https://github.com/microsoft/pyright/blob/7bc2bd8ab05fc4967120d1b6b2a0a1834da2b783/packages/vscode-pyright/src/cancellationUtils.ts#L77 and tell lsp this is the file for cancellation

```ts
export class FileBasedCancellationStrategy implements CancellationStrategy, Disposable {
    constructor() {
        const folderName = randomBytes(21).toString('hex');
        this._sender = new FileCancellationSenderStrategy(folderName);
    }
```

3. which is an argument sent to construct the language server process

```ts
    getCommandLineArguments(): string[] {
        return [`--cancellationReceive=file:${this._sender.folderName}`];
    }
```

and this sender sends cancellation by simply writing empty string to this file with name from `getCancellationFilePath`, typically takes a cancellation id. Assuming we already know from response message [format](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#responseMessage), we know which request to cancel.

```ts
class FileCancellationSenderStrategy implements CancellationSenderStrategy {
    sendCancellation(_: MessageConnection, id: CancellationId): void {
        const file = getCancellationFilePath(this.folderName, id);
        tryRun(() => fs.writeFileSync(file, '', { flag: 'w' }));
    }
```

related discusstion
https://github.com/emacs-lsp/lsp-pyright/issues/13#issuecomment-721511055

**Comment by yyoncho** [comment, marked as answer]
We don't support it, it is not part of the spec (at least I am not aware of it).

**Reply by braineo** [reply]
Thank you. Then it seemed Microsoft expended their own lsp protocol internal.
